File: pt5/1202.smallest-string-with-swaps.java
import java.util.*;

class Solution {
    public String smallestStringWithSwaps(String s, List<List<Integer>> pairs) {
        DSU dsu = new DSU(s.length());
        for (List<Integer> pair : pairs) {
            dsu.union(pair.get(0), pair.get(1));
        }

        Map<Integer, PriorityQueue<Character>> groups = new HashMap<>();
        for (int i = 0; i < s.length(); i++) {
            int root = dsu.find(i);
            groups.computeIfAbsent(root, key -> new PriorityQueue<>()).offer(s.charAt(i));
        }

        StringBuilder answer = new StringBuilder();
        for (int i = 0; i < s.length(); i++) {
            answer.append(groups.get(dsu.find(i)).poll());
        }
        return answer.toString();
    }

    private static class DSU {
        int[] parent;
        int[] rank;

        DSU(int n) {
            parent = new int[n];
            rank = new int[n];
            for (int i = 0; i < n; i++) {
                parent[i] = i;
            }
        }

        int find(int x) {
            if (parent[x] != x) {
                parent[x] = find(parent[x]);
            }
            return parent[x];
        }

        void union(int a, int b) {
            int rootA = find(a);
            int rootB = find(b);
            if (rootA == rootB) {
                return;
            }
            if (rank[rootA] < rank[rootB]) {
                int temp = rootA;
                rootA = rootB;
                rootB = temp;
            }
            parent[rootB] = rootA;
            if (rank[rootA] == rank[rootB]) {
                rank[rootA]++;
            }
        }
    }
}

File: pt5/1202.smallest-string-with-swaps.py
from __future__ import annotations

from collections import defaultdict
from typing import DefaultDict, List


class Solution:
    def smallestStringWithSwaps(self, s: str, pairs: List[List[int]]) -> str:
        parent = list(range(len(s)))
        rank = [0] * len(s)

        def find(x: int) -> int:
            while parent[x] != x:
                parent[x] = parent[parent[x]]
                x = parent[x]
            return x

        def union(a: int, b: int) -> None:
            root_a = find(a)
            root_b = find(b)
            if root_a == root_b:
                return
            if rank[root_a] < rank[root_b]:
                root_a, root_b = root_b, root_a
            parent[root_b] = root_a
            if rank[root_a] == rank[root_b]:
                rank[root_a] += 1

        for a, b in pairs:
            union(a, b)

        characters: DefaultDict[int, List[str]] = defaultdict(list)
        for index, char in enumerate(s):
            characters[find(index)].append(char)

        for group in characters.values():
            group.sort(reverse=True)

        result = []
        for index in range(len(s)):
            result.append(characters[find(index)].pop())

        return "".join(result)

